okapi-core/src/main/java/org/folio/okapi/util/ModuleCache.java

```diff
@@ -0,0 +1,203 @@
+package org.folio.okapi.util;
+
+import io.vertx.core.http.HttpMethod;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.apache.logging.log4j.Logger;
+import org.folio.okapi.bean.ModuleDescriptor;
+import org.folio.okapi.bean.ModuleInstance;
+import org.folio.okapi.bean.RoutingEntry;
+import org.folio.okapi.common.Messages;
+import org.folio.okapi.common.OkapiLogger;
+
+public class ModuleCache {
+  private static final Logger logger = OkapiLogger.get();
+  private static final Messages messages = Messages.getInstance();
+
+  static class ModuleCacheEntry {
+    final ModuleDescriptor moduleDescriptor;
+    final RoutingEntry routingEntry;
+
+    public ModuleCacheEntry(ModuleDescriptor moduleDescriptor, RoutingEntry routingEntry) {
+      this.moduleDescriptor = moduleDescriptor;
+      this.routingEntry = routingEntry;
+    }
+  }
+
+  final Map<String, List<ModuleCacheEntry>> proxyMap = new HashMap<>();
+  final Map<String, List<ModuleCacheEntry>> multiMap = new HashMap<>();
+  final Map<String, List<ModuleCacheEntry>> filterMap = new HashMap<>();
+  final List<ModuleDescriptor> moduleDescriptors;
+
+  /**
+   * Construct cache with module descriptors.
+   * @param moduleDescriptors to be cached
+   */
+  public ModuleCache(List<ModuleDescriptor> moduleDescriptors) {
+    this.moduleDescriptors = moduleDescriptors;
+    for (ModuleDescriptor moduleDescriptor : moduleDescriptors) {
+      add(moduleDescriptor);
+    }
+  }
+
+  /**
+   * Return modules descriptors in cache.
+   * @return list of modules
+   */
+  public List<ModuleDescriptor> getModules() {
+    return moduleDescriptors;
+  }
+
+  static String getPatternPrefix(RoutingEntry re) {
+    String pathPattern = re.getPathPattern();
+    if (pathPattern == null) {
+      return "/"; // anything but pathPattern is legacy so we don't care about those
+    }
+    int index = 0;
+    while (index < pathPattern.length()) {
+      if (pathPattern.charAt(index) == '*' || pathPattern.charAt(index) == '{') {
+        while (index > 0 && pathPattern.charAt(index - 1) != '/') {
+          --index;
+        }
+        break;
+      }
+      index++;
+    }
+    return pathPattern.substring(0, index);
```

Comment:
```suggestion
    int lastSlash = 0;
    for (int i = 0; i < pathPattern.length(); i++) {
      switch (pathPattern.charAt(i)) {
        case '*':
        case '{':
          return pathPattern.substring(0, lastSlash);
        case '/':
          lastSlash = i + 1;
          break;
        default:
          break;
      }
    }
    return pathPattern;
```

Reply:
Ok. A little more lines, but perhaps more readable. In particular it avoids substring when there's no masking.